src/S3/S3Transfer/MultipartUploader.php

```diff
@@ -0,0 +1,542 @@
+<?php
+namespace Aws\S3\S3Transfer;
+
+use Aws\CommandInterface;
+use Aws\CommandPool;
+use Aws\HashingStream;
+use Aws\PhpHash;
+use Aws\ResultInterface;
+use Aws\S3\S3ClientInterface;
+use Aws\S3\S3Transfer\Models\UploadResponse;
+use Aws\S3\S3Transfer\Progress\TransferListener;
+use Aws\S3\S3Transfer\Progress\TransferListenerNotifier;
+use Aws\S3\S3Transfer\Progress\TransferProgressSnapshot;
+use GuzzleHttp\Promise\Coroutine;
+use GuzzleHttp\Promise\Create;
+use GuzzleHttp\Promise\PromiseInterface;
+use GuzzleHttp\Promise\PromisorInterface;
+use GuzzleHttp\Psr7\LazyOpenStream;
+use GuzzleHttp\Psr7\Utils;
+use Psr\Http\Message\StreamInterface;
+use Psr\Http\Message\StreamInterface as Stream;
+use Throwable;
+
+/**
+ * Multipart uploader implementation.
+ */
+class MultipartUploader implements PromisorInterface
+{
+    public const PART_MIN_SIZE = 5 * 1024 * 1024; // 5 MBs
+    public const PART_MAX_SIZE = 5 * 1024 * 1024 * 1024; // 5 GBs
+    public const PART_MAX_NUM = 10000;
+
+    /** @var S3ClientInterface */
+    private readonly S3ClientInterface $s3Client;
+
+    /** @var array */
+    private readonly array $createMultipartArgs;
+
+    /** @var array */
+    private readonly array $config;
+
+    /** @var string | null */
+    private string | null $uploadId;
+
+    /** @var array */
+    private array $parts;
+
+    /** @var StreamInterface */
+    private StreamInterface $body;
+
+    /** @var int */
+    private int $calculatedObjectSize;
+
+    /** @var array */
+    private array $deferFns = [];
```

Comment:
is this property necessary?

Reply:
I used it for making sure some streams when necessary are closed and are not left dangling.